app/services/theme_settings_migrations_runner.rb

```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+class ThemeSettingsMigrationsRunner
+  Migration =
+    Struct.new(:version, :name, :code, :theme_field_id) do
+      def original_name
+        "#{version.to_s.rjust(4, "0")}-#{name}"
+      end
+    end
+
+  MIGRATION_ENTRY_POINT_JS = <<~JS
+    const migrate = require("discourse/theme/migration")?.default;
+    function main(settingsObj) {
+      if (!migrate) {
+        throw new Error("no_exported_migration_function");
+      }
+      if (typeof migrate !== "function") {
+        throw new Error("default_export_is_not_a_function");
+      }
+      const map = new Map(Object.entries(settingsObj));
+      const updatedMap = migrate(map);
+      if (!updatedMap) {
+        throw new Error("migration_function_no_returned_value");
+      }
+      if (!(updatedMap instanceof Map)) {
+        throw new Error("migration_function_wrong_return_type");
+      }
+      return Object.fromEntries(updatedMap.entries());
+    }
+  JS
+
+  private_constant :Migration, :MIGRATION_ENTRY_POINT_JS
+
+  def initialize(theme, limit: 100, timeout: 100, memory: 100.megabytes)
+    @theme = theme
+    @limit = limit
+    @timeout = timeout
+    @memory = memory
+  end
+
+  def run
+    fields = lookup_pending_migrations_fields
+
+    count = fields.count
+    return [] if count == 0
+
+    raise_error("themes.import_error.migrations.too_many_pending_migrations") if count > @limit
```

Comment:
While we are unlikely to hit this error, I'm wondering why we need this because we can just run the migrations in batches under the hood to stay within certain resource limits and that can be transparent to the end user.

Reply:
What do you mean by running the migrations in batches? Do you mean using a single Mini Racer context to run all pending migrations? I think it will be very tricky to do that because safely because we need to calculate a diff between each migration output and if we run all migrations in a single context, the diff will have to be calculated within the context in JavaScript and it will open up an abuse/security vector where our diff calculation logic can be sabotaged by a malicious migration.

Reply:
Sorry let’s rewind this first. What is the reason for imposing a limit on the number of migrations we can run each time per theme?

Reply:
O does this mean the installing and updating of themes eat up one unicorn worker while it is processsign?

Reply:
Yup, that's correct. At some point we should move the theme update process into a sidekiq job, that will be quite a big and complex refactoring.

Reply:
Ahhh ic. The limit makes complete sense to me know. Thank you for clarifying :+1: